File: window_monitor.cpp
#include "window_monitor.h"
#include <psapi.h>
#include <tlhelp32.h>
#include <algorithm>
#include <cctype>

#pragma comment(lib, "psapi.lib")

WindowMonitor::WindowMonitor() {
    screen_width_ = GetSystemMetrics(SM_CXSCREEN);
    screen_height_ = GetSystemMetrics(SM_CYSCREEN);
    fullscreen_threshold_ = 0.95;  // 95%以上视为接近全屏
}

std::optional<WindowInfo> WindowMonitor::GetForegroundWindowInfo() {
    try {
        // 获取前台窗口句柄
        HWND hwnd = GetForegroundWindow();
        if (!hwnd) {
            return std::nullopt;
        }
        
        // 获取窗口标题
        wchar_t window_title_w[256] = {0};
        GetWindowTextW(hwnd, window_title_w, sizeof(window_title_w) / sizeof(wchar_t));
        std::string window_title = WideToUtf8(window_title_w);
        
        // 获取进程ID
        DWORD process_id = 0;
        GetWindowThreadProcessId(hwnd, &process_id);
        if (process_id == 0) {
            return std::nullopt;
        }
        
        // 获取进程名
        std::string process_name = GetProcessName(process_id);
        if (process_name.empty()) {
            process_name = "Process_" + std::to_string(process_id);
        }
        
        // 获取可执行文件路径（可选）
        std::optional<std::string> executable_path = GetExecutablePath(process_id);
        
        // 检查是否接近全屏
        bool is_near_fullscreen = CheckNearFullscreen(hwnd);
        
        WindowInfo info;
        info.process_name = process_name;
        info.window_title = window_title;
        info.is_near_fullscreen = is_near_fullscreen;
        info.process_id = process_id;
        info.executable_path = executable_path;
        
        return info;
    } catch (...) {
        return std::nullopt;
    }
}

bool WindowMonitor::CheckNearFullscreen(HWND hwnd) {
    try {
        RECT rect;
        if (!GetWindowRect(hwnd, &rect)) {
            return false;
        }
        
        int window_width = rect.right - rect.left;
        int window_height = rect.bottom - rect.top;
        
        // 计算窗口占屏幕的比例
        double width_ratio = static_cast<double>(window_width) / screen_width_;
        double height_ratio = static_cast<double>(window_height) / screen_height_;
        
        // 如果宽度和高度都超过阈值，则认为接近全屏
        return width_ratio >= fullscreen_threshold_ && height_ratio >= fullscreen_threshold_;
    } catch (...) {
        return false;
    }
}

std::string WindowMonitor::GetProcessName(DWORD process_id) {
    // 方法1: 使用QueryFullProcessImageName（Windows Vista+）
    HANDLE hProcess = OpenProcess(PROCESS_QUERY_INFORMATION | PROCESS_VM_READ, FALSE, process_id);
    if (hProcess != NULL) {
        wchar_t process_path[MAX_PATH] = {0};
        DWORD size = MAX_PATH;
        
        // 尝试使用QueryFullProcessImageNameW
        if (QueryFullProcessImageNameW(hProcess, 0, process_path, &size)) {
            CloseHandle(hProcess);
            std::string path = WideToUtf8(process_path);
            // 提取文件名
            size_t last_slash = path.find_last_of("\\/");
            if (last_slash != std::string::npos) {
                return path.substr(last_slash + 1);
            }
            return path;
        }
        CloseHandle(hProcess);
    }
    
    // 方法2: 使用CreateToolhelp32Snapshot作为后备
    HANDLE hSnapshot = CreateToolhelp32Snapshot(TH32CS_SNAPPROCESS, 0);
    if (hSnapshot != INVALID_HANDLE_VALUE) {
        PROCESSENTRY32W pe32;
        pe32.dwSize = sizeof(PROCESSENTRY32W);
        
        if (Process32FirstW(hSnapshot, &pe32)) {
            do {
                if (pe32.th32ProcessID == process_id) {
                    CloseHandle(hSnapshot);
                    return WideToUtf8(pe32.szExeFile);
                }
            } while (Process32NextW(hSnapshot, &pe32));
        }
        CloseHandle(hSnapshot);
    }
    
    return "";
}

std::optional<std::string> WindowMonitor::GetExecutablePath(DWORD process_id) {
    HANDLE hProcess = OpenProcess(PROCESS_QUERY_INFORMATION | PROCESS_VM_READ, FALSE, process_id);
    if (hProcess == NULL) {
        return std::nullopt;
    }
    
    wchar_t process_path[MAX_PATH] = {0};
    DWORD size = MAX_PATH;
    
    if (QueryFullProcessImageNameW(hProcess, 0, process_path, &size)) {
        CloseHandle(hProcess);
        return WideToUtf8(process_path);
    }
    
    CloseHandle(hProcess);
    return std::nullopt;
}

std::string WindowMonitor::WideToUtf8(const std::wstring& wstr) {
    if (wstr.empty()) {
        return "";
    }
    
    int size_needed = WideCharToMultiByte(CP_UTF8, 0, wstr.c_str(), -1, NULL, 0, NULL, NULL);
    if (size_needed <= 0) {
        return "";
    }
    
    std::string result(size_needed, 0);
    WideCharToMultiByte(CP_UTF8, 0, wstr.c_str(), -1, &result[0], size_needed, NULL, NULL);
    
    // 移除末尾的null字符
    if (!result.empty() && result.back() == '\0') {
        result.pop_back();
    }
    
    return result;
}

std::string WindowMonitor::ToLower(const std::string& str) {
    std::string result = str;
    std::transform(result.begin(), result.end(), result.begin(),
                   [](unsigned char c) { return std::tolower(c); });
    return result;
}

File: window_monitor.py
"""
窗口监控模块 - 获取前台窗口信息
"""
import win32gui
import win32process
import win32api
import win32con
import psutil
from typing import Optional, Dict, Tuple
from dataclasses import dataclass


@dataclass
class WindowInfo:
    """窗口信息数据类"""
    process_name: str  # 可执行文件名
    window_title: str  # 窗口标题
    is_near_fullscreen: bool  # 是否接近全屏
    process_id: int  # 进程ID
    executable_path: Optional[str] = None  # 可执行文件路径


class WindowMonitor:
    """窗口监控器 - 负责获取前台窗口信息"""
    
    def __init__(self):
        self._screen_width = win32api.GetSystemMetrics(win32con.SM_CXSCREEN)
        self._screen_height = win32api.GetSystemMetrics(win32con.SM_CYSCREEN)
        self._fullscreen_threshold = 0.95  # 95%以上视为接近全屏
    
    def get_foreground_window_info(self) -> Optional[WindowInfo]:
        """
        获取当前前台窗口信息
        
        Returns:
            WindowInfo对象，如果获取失败则返回None
        """
        try:
            # 获取前台窗口句柄
            hwnd = win32gui.GetForegroundWindow()
            if not hwnd:
                return None
            
            # 获取窗口标题
            window_title = win32gui.GetWindowText(hwnd)
            
            # 获取进程ID
            _, process_id = win32process.GetWindowThreadProcessId(hwnd)
            
            # 获取进程信息
            process_name = None
            executable_path = None
            try:
                process = psutil.Process(process_id)
                process_name = process.name()
                try:
                    executable_path = process.exe()
                except (psutil.AccessDenied, psutil.NoSuchProcess):
                    executable_path = None
            except (psutil.NoSuchProcess, psutil.AccessDenied):
                # 如果无法获取进程信息，使用进程ID作为后备
                process_name = f"Process_{process_id}"
                executable_path = None
            
            # 检查是否接近全屏
            is_near_fullscreen = self._check_near_fullscreen(hwnd)
            
            return WindowInfo(
                process_name=process_name,
                window_title=window_title,
                is_near_fullscreen=is_near_fullscreen,
                process_id=process_id,
                executable_path=executable_path
            )
        except Exception as e:
            print(f"获取窗口信息时出错: {e}")
            return None
    
    def _check_near_fullscreen(self, hwnd: int) -> bool:
        """
        检查窗口是否接近全屏
        
        Args:
            hwnd: 窗口句柄
            
        Returns:
            如果窗口占据屏幕95%以上则返回True
        """
        try:
            # 获取窗口矩形
            rect = win32gui.GetWindowRect(hwnd)
            left, top, right, bottom = rect
            
            # 计算窗口大小
            window_width = right - left
            window_height = bottom - top
            
            # 计算窗口占屏幕的比例
            width_ratio = window_width / self._screen_width
            height_ratio = window_height / self._screen_height
            
            # 如果宽度和高度都超过阈值，则认为接近全屏
            return width_ratio >= self._fullscreen_threshold and height_ratio >= self._fullscreen_threshold
        except Exception:
            return False

